[data views] Allow data views created on hidden and system indices - second attempt

mattkime opened the pull request:
## Summary

Previously, the 'Allow hidden and system indices' advanced option when creating a data view was only a UI convenience. It allowed you to see which hidden and system indices you were matching but they would be would be selected just the same once the data view was loaded. At some point something changed and now there are system and hidden indices that require `expandWildcards: hidden` to be passed to field caps in order to see anything. `allowHidden: boolean` is added to the DataView and DataViewSpec and passed through when field caps requests are made.

This is primarily a tool for troubleshooting. For instance, instead of hitting a full data stream across a number of data tiers you can select a specific index to compare its performance.

NOTE: This is a second attempt. What I learned - the whole `expand_wildcards` param is literal - you can directly query a hidden index without `expandWildcards: hidden` since its not using a wildcard. Tests now use a wildcard.

Closes: https://github.com/elastic/kibana/issues/164652

elasticmachine commented:
Pinging @elastic/kibana-data-discovery (Team:DataDiscovery)

davismcphee reviewed (commented):
The changes look good overall, but I encountered a couple of things worth looking into while testing.

After editing a data view targeting a hidden index, if I reopen the edit flyout without refreshing the page, the "Allow hidden and system indices" toggle remains unchecked. Once I reload the page and open the flyout, the toggle is checked as expected:

https://github.com/elastic/kibana/assets/25592674/56d60ec6-88b2-4e9a-a10c-75f4553c0ae1

Whenever I target a hidden index using a pattern that doesn't start with a `.` and includes a `*`, I get 404 responses in the network tab even though the pattern should match. The time field input also becomes disabled when this happens, preventing me from selecting a time field. If I instead target the same index with an exact pattern, things work as expected:

https://github.com/elastic/kibana/assets/25592674/205fb405-e68d-423b-9156-11323a8b38d9

Also for some reason when I save the data view with the problematic pattern, I get no results in Discover, but I get results when using the data view with the exact pattern:

https://github.com/elastic/kibana/assets/25592674/825f30f9-6b9c-4416-bc94-530411e41d8d

mattkime commented:
@davismcphee I believe I've addressed the issues you saw although I had trouble reproducing the 3rd video. 

The problem was `allowHidden` wasn't being passed to loading the field list. The functional test now specifies the timestamp field name which should prevent this from happening again in the future.

davismcphee reviewed (commented):
Tested again and it seems like the first two issues are fixed 👍 I'm still encountering the third issue, but I'm starting to think it's either a Discover or search issue. Let's sync next week and go through it together, and if it seems unrelated we can merge this PR and open a separate issue.

mattkime commented:
@davismcphee I was able to reproduce the problem you captured where a query works against a single hidden index but not a set of hidden indices. I think we need to ask for @lukasolson 's advice on how `expand_wildcards` might be addressed on the search side of things.

lukasolson reviewed (commented):
I don't think we've ever supported searching hidden indices (other than system indices, which it seems don't need the `expand_wildcard` option). If you wanted to add support, it would probably look something like this:


```
diff --git a/src/plugins/data/common/search/search_source/fetch/get_search_params.ts b/src/plugins/data/common/search/search_source/fetch/get_search_params.ts
index 5f8ca973b3d..9be4e2f6208 100644
--- a/src/plugins/data/common/search/search_source/fetch/get_search_params.ts
+++ b/src/plugins/data/common/search/search_source/fetch/get_search_params.ts
@@ -44,6 +44,7 @@ export function getSearchParamsFromRequest(
     body,
     // @ts-expect-error `track_total_hits` not allowed at top level for `typesWithBodyKey`
     track_total_hits,
+    ...(searchRequest.index?.allowHidden && { expand_wildcards: 'all' }),
     ...searchParams,
   };
 }
```

We would want to add a test case for this as well. If you'd rather defer this change to a separate PR, that might also make sense.

mbondyra reviewed (approved):
LGTM 👌🏼

davismcphee reviewed (approved):
Thanks both for looking into and addressing the wildcard issue! I pulled and tested again and can confirm the issue no longer exists. It seems hidden and system index data views are now fully supported, LGTM 👍

yngrdyn reviewed (approved):
LGTM

kertal commented:
@elasticmachine merge upstream

yctercero commented:
The exceptions flake occurring seems to be completely unrelated to these changes. It's expecting one rule to show in the test instead of the two appearing for selection. Skipping and tracking in this [PR](https://github.com/elastic/kibana/pull/171236) here.
